ceph/ceph.go

```diff
@@ -83,98 +85,176 @@ type RealCmd struct{}
 		return nil, err
 	}
 
-	metrics := []plugin.PluginMetricType{}
+	daemonNameSplit := strings.Split(trimPrefixAndSuffix(daemon, ceph.socket.prefix, "."+ceph.socket.ext), ".")
+	daemonName := daemonNameSplit[0]
+	daemonID := daemonNameSplit[1]
+
+	metrics := []plugin.MetricType{}
 
 	for _, m := range mts {
-		nlen := len(m.Namespace())
-		daemonName := trimPrefixAndSuffix(daemon, c.socket.prefix, "."+c.socket.ext)
-
-		// compare daemonName with proper component of metric's namespace [intel storage ceph daemonName]
-		if daemonName == m.Namespace()[daemonNameIndex] {
-			// get metrics defined in task for this daemon
-			dat_r := dat
-			// skip the const components of metrics namespace
-			for _, name := range m.Namespace()[daemonNameIndex+1 : nlen-1] {
-				if dat_r[name] == nil {
-					break
-				}
-				dat_r = dat_r[name].(map[string]interface{}) // get metric
+		// Get metrics defined in task for this daemon
+		if matchSlice(m.Namespace().Strings()[daemonNameIndex:daemonIDIndex+1], daemonNameSplit) {
+			daemonMetrics := make(map[string]interface{})
+			ceph.getJSONDataByNamespace(jsonData, m.Namespace().Strings()[daemonIDIndex+1:], []string{}, &daemonMetrics)
+
+			// No metrics found for desired namespace
+			if len(daemonMetrics) == 0 {
+				daemonMetrics[strings.Join(m.Namespace().Strings()[daemonIDIndex+1:], "/")] = nil
 			}
 
-			hostname, _ := os.Hostname()
-			metric := plugin.PluginMetricType{
-				Namespace_: m.Namespace(),
-				Data_:      dat_r[m.Namespace()[nlen-1]], // get value of metric
-				Source_:    hostname + "/" + daemonName,
-				Timestamp_: timestamp,
-			}
+			for ns, data := range daemonMetrics {
+				metric := plugin.MetricType{
+					Namespace_: core.NewNamespace(m.Namespace().Strings()[:daemonIDIndex]...).AddStaticElement(daemonID).AddStaticElements(strings.Split(ns, "/")...),
+					Data_:      data, // get value of metric
+					Tags_:      map[string]string{core.STD_TAG_PLUGIN_RUNNING_ON: hostname + "/" + daemonName + "." + daemonID},
+					Timestamp_: timestamp,
+				}
 
-			metrics = append(metrics, metric)
+				metrics = append(metrics, metric)
+			}
 		}
 	}
 
-	if len(metrics) <= 0 {
-		// do not find desired metrics
-		return nil, errors.New("Do not find desired metrics for ceph-deamon")
+	// No metrics fount at all
+	if len(metrics) == 0 {
+		return metrics, errors.New("No metrics found!")
 	}
 
 	return metrics, nil
 }
 
-// CollectMetrics returns all desired Ceph metrics defined in task manifest
-func (ceph *Ceph) CollectMetrics(mts []plugin.PluginMetricType) ([]plugin.PluginMetricType, error) {
+// matchSlice matches 2 slices with asterisk support
+func matchSlice(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	for i := range a {
+		if a[i] != b[i] && a[i] != "*" && b[i] != "*" {
+			return false
+		}
+	}
 
+	return true
+}
+
+func (ceph *Ceph) getJSONDataByNamespace(data map[string]interface{}, namespace []string, resultNamespace []string, results *map[string]interface{}) {
```

Comment:
This function works slightly different that the one from snap-plugin-utilities and is designed for this specific case. It breaks JSON keys at dots, for example mutex-OSD:ShardedOpWQ:.0 is divided into mutex-OSD:ShardedOpWQ:/0 namespace, which is then compared key-by-key with namespace provided in parameter with asterisk matching.